File: apps/desktop/electron/db/mongo.ts
import { MongoClient, Db } from 'mongodb'
import { DatabaseConnection, ConnectionConfig, StorageAnalysis, QueryResult, SchemaInfo } from './index'
import { writeFileSync } from 'fs'

export class MongoDBConnection extends DatabaseConnection {
  private client?: MongoClient
  private db?: Db

  async connect(): Promise<void> {
    const connectionString = this.config.connectionString || 
      `mongodb://${this.config.username ? `${this.config.username}:${this.config.password}@` : ''}${this.config.host}:${this.config.port || 27017}/${this.config.database || ''}`

    this.client = new MongoClient(connectionString)
    await this.client.connect()
    
    this.db = this.client.db(this.config.database)
    this.connected = true
  }

  async disconnect(): Promise<void> {
    if (this.client) {
      await this.client.close()
      this.client = undefined
      this.db = undefined
      this.connected = false
    }
  }

  async testConnection(): Promise<void> {
    if (!this.client) {
      await this.connect()
    }
    
    await this.db!.admin().ping()
  }

  async analyzeStorage(): Promise<StorageAnalysis> {
    if (!this.client) {
      await this.connect()
    }

    const db = this.db!

    // Get database stats
    const dbStats = await db.stats()
    const totalSize = dbStats.dataSize + dbStats.indexSize

    // Get collections
    const collections = await db.listCollections().toArray()
    const tables = []

    for (const collection of collections) {
      const collectionName = collection.name
      const coll = db.collection(collectionName)
      
      // Get collection stats
      const collStats = await db.command({ collStats: collectionName })
      const rowCount = collStats.count || 0
      const size = collStats.size || 0
      const indexSize = collStats.totalIndexSize || 0

      tables.push({
        name: collectionName,
        size,
        rowCount,
        indexSize,
        bloat: 0 // MongoDB doesn't have traditional bloat
      })
    }

    // Sort by size descending
    tables.sort((a, b) => b.size - a.size)

    // Get indexes
    const indexes = []
    for (const collection of collections) {
      const collectionName = collection.name
      const coll = db.collection(collectionName)
      const indexList = await coll.indexes()
      
      for (const index of indexList) {
        // Get index size (approximate)
        const indexStats = await db.command({ collStats: collectionName })
        const avgIndexSize = indexStats.totalIndexSize / indexList.length

        indexes.push({
          name: index.name,
          tableName: collectionName,
          size: avgIndexSize,
          bloat: 0
        })
      }
    }

    const largestTable = tables.length > 0 ? tables[0] : { name: 'N/A', size: 0, rowCount: 0 }

    return {
      totalSize,
      tableCount: tables.length,
      indexCount: indexes.length,
      largestTable,
      tables,
      indexes,
      lastAnalyzed: new Date().toISOString()
    }
  }

  async executeQuery(query: string): Promise<QueryResult> {
    if (!this.client) {
      await this.connect()
    }

    const db = this.db!
    const startTime = Date.now()

    try {
      // Parse MongoDB query (simplified - in real implementation, you'd want a proper parser)
      const queryObj = JSON.parse(query)
      const { collection, operation, filter, projection, sort, limit } = queryObj

      if (!collection || !operation) {
        throw new Error('Query must specify collection and operation')
      }

      const coll = db.collection(collection)
      let result: any[] = []

      switch (operation.toLowerCase()) {
        case 'find':
          const cursor = coll.find(filter || {}, { projection })
          if (sort) cursor.sort(sort)
          if (limit) cursor.limit(limit)
          result = await cursor.toArray()
          break
        case 'aggregate':
          result = await coll.aggregate(queryObj.pipeline || []).toArray()
          break
        case 'count':
          result = [{ count: await coll.countDocuments(filter || {}) }]
          break
        default:
          throw new Error(`Unsupported operation: ${operation}`)
      }

      const executionTime = Date.now() - startTime

      // Get column names from the first document
      const columns = result.length > 0 ? Object.keys(result[0]) : []

      return {
        columns,
        rows: result,
        rowCount: result.length,
        executionTime,
        explainPlan: undefined // MongoDB explain is more complex
      }
    } catch (error) {
      throw new Error(`Query execution failed: ${error instanceof Error ? error.message : 'Unknown error'}`)
    }
  }

  async getSchema(): Promise<SchemaInfo> {
    if (!this.client) {
      await this.connect()
    }

    const db = this.db!

    // Get collections
    const collections = await db.listCollections().toArray()
    const tables = []

    for (const collection of collections) {
      const collectionName = collection.name
      const coll = db.collection(collectionName)

      // Get sample document to infer schema
      const sampleDoc = await coll.findOne({})
      const columns = sampleDoc ? Object.keys(sampleDoc).map(key => ({
        name: key,
        type: typeof sampleDoc[key],
        nullable: true, // MongoDB fields are always nullable
        defaultValue: undefined
      })) : []

      // Get indexes
      const indexList = await coll.indexes()
      const indexes = indexList.map(index => ({
        name: index.name,
        columns: Object.keys(index.key),
        unique: index.unique || false
      }))

      tables.push({
        name: collectionName,
        columns,
        indexes
      })
    }

    return {
      schemas: [{
        name: this.config.database || 'default',
        tables
      }]
    }
  }

  async createBackup(backupPath: string): Promise<{ path: string; size: number }> {
    if (!this.client) {
      await this.connect()
    }

    const db = this.db!

    try {
      // Get all collections and their data
      const collections = await db.listCollections().toArray()
      const backupData: any = {
        database: this.config.database,
        collections: {}
      }

      for (const collection of collections) {
        const collectionName = collection.name
        const coll = db.collection(collectionName)
        const data = await coll.find({}).toArray()
        backupData.collections[collectionName] = data
      }

      // Write backup to file
      const backupJson = JSON.stringify(backupData, null, 2)
      writeFileSync(backupPath, backupJson)

      const stats = require('fs').statSync(backupPath)
      return { path: backupPath, size: stats.size }
    } catch (error) {
      throw new Error(`Failed to create MongoDB backup: ${error instanceof Error ? error.message : 'Unknown error'}`)
    }
  }

  async restoreBackup(backupPath: string): Promise<void> {
    if (!this.client) {
      await this.connect()
    }

    const db = this.db!

    try {
      // Read backup file
      const backupData = JSON.parse(require('fs').readFileSync(backupPath, 'utf8'))

      // Clear existing collections
      const collections = await db.listCollections().toArray()
      for (const collection of collections) {
        await db.collection(collection.name).drop()
      }

      // Restore collections
      for (const [collectionName, data] of Object.entries(backupData.collections)) {
        if (Array.isArray(data) && data.length > 0) {
          await db.collection(collectionName).insertMany(data as any[])
        }
      }
    } catch (error) {
      throw new Error(`Failed to restore MongoDB backup: ${error instanceof Error ? error.message : 'Unknown error'}`)
    }
  }
}

File: db_storage_manager/db/mongo.py
"""
MongoDB database connection
"""

import asyncio
import subprocess
import tarfile
import tempfile
from pathlib import Path
from typing import Any, Dict, List
from pymongo import MongoClient
from pymongo.errors import ConnectionFailure

from .base import (
    DatabaseConnection,
    ConnectionConfig,
    StorageAnalysis,
    QueryResult,
    SchemaInfo,
)


class MongoDBConnection(DatabaseConnection):
    """MongoDB database connection"""

    def __init__(self, config: ConnectionConfig):
        super().__init__(config)
        self.client = None
        self.db = None

    async def connect(self) -> None:
        """Connect to MongoDB database"""
        if self.config.connection_string:
            self.client = MongoClient(self.config.connection_string)
        else:
            connection_uri = f"mongodb://{self.config.username}:{self.config.password}@{self.config.host or 'localhost'}:{self.config.port or 27017}/{self.config.database}"
            self.client = MongoClient(connection_uri)

        # Test connection
        self.client.admin.command("ping")
        self.db = self.client[self.config.database or "admin"]
        self.connected = True

    async def disconnect(self) -> None:
        """Disconnect from MongoDB database"""
        if self.client:
            self.client.close()
            self.client = None
            self.db = None
        self.connected = False

    async def analyze_storage(self) -> StorageAnalysis:
        """Analyze MongoDB storage"""
        if not self.db:
            await self.connect()

        collections = self.db.list_collection_names()
        tables = []
        total_size = 0

        for collection_name in collections:
            collection = self.db[collection_name]
            stats = self.db.command("collStats", collection_name)

            table_size = stats.get("size", 0)
            index_size = stats.get("totalIndexSize", 0)
            row_count = stats.get("count", 0)

            tables.append(
                {
                    "name": collection_name,
                    "size": table_size,
                    "rowCount": row_count,
                    "indexSize": index_size,
                    "bloat": 0.0,
                }
            )
            total_size += table_size + index_size

        # Sort by size
        tables.sort(key=lambda x: x["size"], reverse=True)

        # Get indexes
        indexes = []
        for collection_name in collections:
            collection = self.db[collection_name]
            index_list = collection.list_indexes()

            for index in index_list:
                indexes.append(
                    {
                        "name": index.get("name", "unnamed"),
                        "tableName": collection_name,
                        "size": 0,
                        "bloat": 0.0,
                    }
                )

        largest_table = (
            tables[0] if tables else {"name": "N/A", "size": 0, "rowCount": 0}
        )

        return {
            "totalSize": total_size,
            "tableCount": len(tables),
            "indexCount": len(indexes),
            "largestTable": largest_table,
            "tables": tables,
            "indexes": indexes,
            "lastAnalyzed": __import__("datetime").datetime.now().isoformat(),
        }

    async def execute_query(self, query: str, safe_mode: bool = True) -> QueryResult:
        """Execute a MongoDB query"""
        if not self.db:
            await self.connect()

        import time
        import json

        start_time = time.time()

        try:
            # Parse query as JSON
            query_dict = json.loads(query)
            collection_name = query_dict.get("collection", "default")
            collection = self.db[collection_name]

            if "find" in query_dict:
                result = list(collection.find(query_dict["find"]))
            elif "aggregate" in query_dict:
                result = list(collection.aggregate(query_dict["aggregate"]))
            else:
                # Fallback to direct command
                result = [self.db.command(query_dict)]

            columns = list(result[0].keys()) if result else []
            rows = [dict(item) for item in result]

            execution_time = int((time.time() - start_time) * 1000)

            return {
                "columns": columns,
                "rows": rows,
                "rowCount": len(rows),
                "executionTime": execution_time,
                "explainPlan": None,
            }
        except json.JSONDecodeError:
            # If not JSON, try as direct command
            try:
                result = [self.db.command(query)]
                columns = list(result[0].keys()) if result else []
                rows = [dict(item) for item in result]

                execution_time = int((time.time() - start_time) * 1000)

                return {
                    "columns": columns,
                    "rows": rows,
                    "rowCount": len(rows),
                    "executionTime": execution_time,
                    "explainPlan": None,
                }
            except Exception as e:
                raise RuntimeError(f"Query execution failed: {str(e)}")

    async def get_schema(self) -> SchemaInfo:
        """Get MongoDB schema"""
        if not self.db:
            await self.connect()

        collections = self.db.list_collection_names()
        tables = []

        for collection_name in collections:
            collection = self.db[collection_name]

            # Get sample document to infer schema
            sample_doc = collection.find_one()
            columns = []
            if sample_doc:
                for key, value in sample_doc.items():
                    columns.append(
                        {
                            "name": key,
                            "type": type(value).__name__,
                            "nullable": True,
                            "defaultValue": None,
                        }
                    )

            # Get indexes
            index_list = collection.list_indexes()
            indexes = []
            for index in index_list:
                indexes.append(
                    {
                        "name": index.get("name", "unnamed"),
                        "columns": list(index.get("key", {}).keys()),
                        "unique": index.get("unique", False),
                    }
                )

            tables.append(
                {
                    "name": collection_name,
                    "columns": columns,
                    "indexes": indexes,
                }
            )

        return {
            "schemas": [
                {
                    "name": self.config.database or "default",
                    "tables": tables,
                }
            ]
        }

    async def create_backup(self, backup_path: str) -> Dict[str, Any]:
        """Create MongoDB backup using mongodump"""
        if not self.db:
            await self.connect()

        backup_file = Path(backup_path)
        backup_file.parent.mkdir(parents=True, exist_ok=True)

        # Create temporary directory for dump
        with tempfile.TemporaryDirectory() as temp_dir:
            temp_path = Path(temp_dir)

            # Build mongodump command
            cmd = [
                "mongodump",
                "--host",
                f"{self.config.host or 'localhost'}:{self.config.port or 27017}",
                "--db",
                self.config.database,
                "--username",
                self.config.username,
                "--password",
                self.config.password,
                "--authenticationDatabase",
                "admin",
                "--out",
                str(temp_path),
            ]

            # Run mongodump
            process = await asyncio.create_subprocess_exec(
                *cmd,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )

            stdout, stderr = await process.communicate()

            if process.returncode != 0:
                raise RuntimeError(f"mongodump failed: {stderr.decode()}")

            # Compress dump directory
            with tarfile.open(backup_file, "w:gz") as tar:
                tar.add(temp_path, arcname=".")

        size = backup_file.stat().st_size
        return {"path": str(backup_file), "size": size}

    async def restore_backup(self, backup_path: str) -> None:
        """Restore MongoDB backup using mongorestore"""
        if not self.db:
            await self.connect()

        backup_file = Path(backup_path)
        if not backup_file.exists():
            raise FileNotFoundError(f"Backup file not found: {backup_path}")

        # Create temporary directory for extraction
        with tempfile.TemporaryDirectory() as temp_dir:
            temp_path = Path(temp_dir)

            # Extract backup
            with tarfile.open(backup_file, "r:gz") as tar:
                tar.extractall(temp_path)

            # Build mongorestore command
            cmd = [
                "mongorestore",
                "--host",
                f"{self.config.host or 'localhost'}:{self.config.port or 27017}",
                "--db",
                self.config.database,
                "--username",
                self.config.username,
                "--password",
                self.config.password,
                "--authenticationDatabase",
                "admin",
                str(temp_path),
            ]

            # Run mongorestore
            process = await asyncio.create_subprocess_exec(
                *cmd,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )

            stdout, stderr = await process.communicate()

            if process.returncode != 0:
                raise RuntimeError(f"mongorestore failed: {stderr.decode()}")
